catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/MetacardComparator.java

```diff
@@ -13,106 +13,167 @@
+     * it was last retrieved. Since there is no single attribute that clearly indicates whether
+     * the resource has changed or not, this class compares as many attributes as possible and if
+     * it find a difference, it will return {@code false}.
+     *
+     * @param cachedMetacard  version of the metacard when the product was added to the cache
+     * @param updatedMetacard current version of the metacard
+     * @return {@code true} only if the attributes in the updated metacard do not indicate any
+     * potential changes in the resource associated with it
+     */
+    public static boolean isSame(@Nullable Metacard cachedMetacard,
+            @Nullable Metacard updatedMetacard) {
+
+        if (cachedMetacard == null && updatedMetacard == null) {
+            return true;
+        }
 
-            for (AttributeDescriptor descriptor : cachedDescriptors) {
+        if (cachedMetacard == null || updatedMetacard == null) {
+            return false;
+        }
 
-                attrName = descriptor.getName();
-                cachedAttr = cachedMetacard.getAttribute(attrName);
-                updatedAttr = updatedMetacard.getAttribute(attrName);
+        if (!Objects.equals(cachedMetacard.getId(), updatedMetacard.getId())) {
+            return false;
+        }
 
-                // check if both are null or if not null, check if the values are the same
-                if ((cachedAttr == updatedAttr) || (cachedAttr.toString()
-                        .equals(updatedAttr.toString()))) {
-                    result = true;
-                } else {
-                    result = false;
-                    break;
-                }
+        if (!Objects.equals(cachedMetacard.getAttribute(Metacard.CHECKSUM),
+                updatedMetacard.getAttribute(Metacard.CHECKSUM))) {
+            return false;
+        }
 
-            }
+        if (!allMetacardMethodsReturnMatchingAttributes(cachedMetacard, updatedMetacard)) {
+            return false;
         }
 
-        return result;
+        return allMetacardAttributesEqual(cachedMetacard, updatedMetacard);
     }
 
-    public static boolean isSame(Metacard cachedMetacard, Metacard updatedMetacard) {
-
-        boolean result = false;
-
-        if (cachedMetacard.getId()
-                .equals(updatedMetacard.getId())) {
-
-            Attribute cachedMetacardAttr = cachedMetacard.getAttribute(Metacard.CHECKSUM);
-
-            if (cachedMetacardAttr != null) {
+    private static boolean allMetacardMethodsReturnMatchingAttributes(Metacard cachedMetacard,
+            Metacard updatedMetacard) {
+        Optional<Function<Metacard, ?>> difference = METACARD_METHODS.stream()
+                .filter(method -> !Objects.deepEquals(method.apply(cachedMetacard),
+                        method.apply(updatedMetacard)))
+                .findFirst();
+
+        if (LOGGER.isDebugEnabled() && difference.isPresent()) {
```

Comment:
Good question.

I didn't have any logging statements here initially but they came in handy when I started testing the code so I left them.

My take on this was that it may be useful at runtime to know what causes certain resources to be evicted from the cache and re-downloaded. However, always comparing all attributes felt like a waste since we really only care about the first difference.

I don't have a strong opinion and can change or remove that code if you feel that logging code isn't useful the way it is right now.